packages/dds/quorum/src/quorum.ts

```diff
@@ -0,0 +1,373 @@
+        super(id, runtime, attributes);
+
+        this.incomingOp.on("set", this.handleIncomingSet);
+        this.incomingOp.on("delete", this.handleIncomingDelete);
+        this.incomingOp.on("noop", this.handleIncomingNoOp);
+
+        this.disconnectWatcher.on("disconnect", () => {
+            // assert(this.runtime.clientId !== undefined, 0x1d3 /* "Missing client id on disconnect" */);
+            // TODO Handle appropriately
+        });
+    }
+
+    public has(key: string): boolean {
+        return this.values.get(key)?.accepted !== undefined;
+    }
+
+    public get(key: string): any {
+        return this.values.get(key)?.accepted?.value;
+    }
+
+    // TODO: Should this return differently for a value of undefined vs. a pending delete?
+    public getPending(key: string): any {
+        const pending = this.values.get(key)?.pending;
+        if (pending === undefined || pending.type === "delete") {
+            return undefined;
+        }
+
+        return pending.value;
+    }
+
+    public set(key: string, value: any): void {
+        // TODO: handle detached scenario, just auto accept basically
+
+        const setOp: IQuorumSetOperation = {
+            type: "set",
+            key,
+            value,
+            refSeq: this.runtime.deltaManager.lastSequenceNumber,
+        };
+
+        this.submitLocalMessage(setOp);
+    }
+
+    public delete(key: string): void {
+        // TODO: handle detached scenario, just auto accept basically
+
+        const deleteOp: IQuorumDeleteOperation = {
+            type: "delete",
+            key,
+            refSeq: this.runtime.deltaManager.lastSequenceNumber,
+        };
+
+        this.submitLocalMessage(deleteOp);
+    }
+
+    private readonly handleIncomingSet = (key: string, value: any, refSeq: number, setSequenceNumber: number) => {
+        const currentValue = this.values.get(key);
+        // A proposal is valid if the value is unknown
+        // or if it was made with knowledge of the most recently accepted value
+        const proposalValid =
+            currentValue === undefined
+            || (currentValue.pending === undefined && currentValue.accepted.sequenceNumber <= refSeq);
+        if (!proposalValid) {
+            // Drop invalid proposals on the ground.  If set() returns a promise we will need to resolve it though.
+            return;
+        }
+
+        const accepted = currentValue?.accepted;
+
+        this.values.set(key, { accepted, pending: { type: "set", value, sequenceNumber: setSequenceNumber }});
+
+        this.emit("pending", key);
+
+        // Emit a noop upon a new key entering pending state, which is how we'll eventually advance it to accepted
+        // TODO: Doesn't work if there's a holdout client that disconnects prior to sending noop, since we aren't
+        // guaranteed to get any further message advancing the MSN.  Observing client disconnects could work.
```

Comment:
client disconnects will result in an eventual leave message, but it could take up to 5 minutes, but i don't think you'll ever get perma-stuck.